namerd/storage/zk/src/main/scala/com/twitter/finagle/serverset2/client/buoyant/ZkSession.scala

```diff
@@ -123,71 +124,83 @@ class ZkSession(
       @volatile var closed = false
 
       def loop(): Future[Unit] = {
-        if (!closed) safeRetry(go).respond {
+        if (closed) Future.Unit
+        else safeRetry(go).transform {
           case Throw(e@KeeperException.SessionExpired(_)) =>
             // don't retry. The session has expired while trying to set the watch.
             // In case our activity is still active, notify the listener
             u() = Activity.Failed(e)
+            Future.Unit
 
           case Throw(exc) =>
             logger.error("Operation failed with %s. Session %s", exc, sessionId)
             u() = Activity.Failed(exc)
-            fireAndForget { retryWithDelay { loop() } }
+            retryWithDelay { loop() }
 
           case Return(Watched(value, state)) =>
             val ok = Activity.Ok(value)
             retryBackoff.reset()
             u() = ok
 
-            val _ = state.changes.respond {
-              case WatchState.Pending =>
-              // Ignore updates WatchState is Pending.
+            state.map { s =>
+              s match {
+                case WatchState.Pending => // No action to take
 
-              case WatchState.Determined(_) =>
-                // Note: since the watch transitioned to determined, we know
-                // that this observation will produce no more values, so there's
-                // no need to apply concurrency control to the subsequent
-                // branches.
-                fireAndForget { loop() }
+                case WatchState.Determined(_) => // No action to take
+
+                case WatchState.SessionState(sessionState) if sessionState == SessionState.ConnectedReadOnly |
+                  sessionState == SessionState.SaslAuthenticated |
+                  sessionState == SessionState.SyncConnected =>
+                  logger.info("Reacquiring watch on %s. Session: %s", sessionState, sessionId)
+                  u() = ok
+
+                case WatchState.SessionState(SessionState.Expired) =>
+                  u() = Activity.Failed(new Exception("session expired"))
+
+                case WatchState.SessionState(sessionState) if sessionState == SessionState.Disconnected |
+                  sessionState == SessionState.NoSyncConnected =>
+                  logger.warning(
+                    "Intermediate Failure session state: %s. Session: %s. Data is now unavailable.",
+                    sessionState, sessionId
+                  )
+                  u() = Activity.Failed(new Exception("" + sessionState))
+
+                case WatchState.SessionState(sessionState) =>
+                  logger.error("Unexpected session state %s. Session: %s", sessionState, sessionId)
+                  u() = Activity.Failed(new Exception("" + sessionState))
+              }
+              s
+            }.changes.filter {
```

Comment:
This filter could use a comment explaining that this, in combination with the `toFuture()` below, creates a future that is satisfied when the state becomes Determined, ConnectedReadOnly, SaslAuthenticated, SyncConnected, or an unknown state.